### ISSUE TITLE
Reason is lost in muc:invite events

### ISSUE THREAD

**Issue by kai670** [legastero/stanza#381]
I'm receiving the next message with an invitation:
```
<message id="4fd708bd0eb29b7e" to="dev6@domain" xmlns="jabber:client" from="call-rb8j8seq@conference.domain">
    <x xmlns="http://jabber.org/protocol/muc#user">
        <invite from="admin@domain/QfQz53QX">
            <reason>You've been added to the conference's chat room</reason>
        </invite>
    </x>
    <body>admin@domain/QfQz53QX invited you to the room call-rb8j8seq@conference.quobis (You've been added to the conference's chat room)</body>
    <x xmlns="jabber:x:conference" jid="call-rb8j8seq@domain.quobis">You've been added to the conference's chat room</x>
    <stanza-id id="1dHg6QfzhDdy2oXP" xmlns="urn:xmpp:sid:0" by="dev6@domain" />
</message>
```

And stanza emits a `muc:invite` event with the next object as payload:
```
{
  from: "call-rb8j8seq@conference.quobis"
  password: undefined
  reason: undefined
  room: "call-rb8j8seq@conference.quobis"
  thread: undefined
  type: "direct"
}
```

But I was expecting reason to be: "You've been added to the conference's chat room"

**Comment by legastero** [comment]
Still digging through this one, but for a direct MUC invite (XEP-249), the reason text needs to be in an attribute, not as child text.

So this part of that stanza isn't to spec:

```xml
<x xmlns="jabber:x:conference" jid="call-rb8j8seq@domain.quobis">You've been added to the conference's chat room</x>
```

Your stanza _also_ includes a mediated invite, so I'm checking on how having a mix of invite methods is handled. However, the mediate invite element in your stanza is _also_ incorrect, and missing the room JID.

**Comment by legastero** [comment]
Ahhh, it looks like I'm going to need to tweak the JXT mappings a little bit. The current mapping allows either a mediated or direct invite in one stanza, but not _both_ at once.

So the direct invite is the one that's getting plucked and mapped, and its reason text isn't in the expected location.

**Comment by legastero** [comment]
Oh, I just found a small note in XEP-0045 about this:

> Implementation Note: In the past, it was specified that a <x xmlns='jabber:x:conference'> element with the reason as text payload was to be included in the mediated invitation as sent by the room. While this has since been removed from this specification, implementations should be aware that there still exist server implementations which emit that payload for compatibility reasons.

Ok, so this is going to take some work to properly address.

**Comment by legastero** [comment]
Made a patch in 12.3.2 to at least get the reason text pulled in this situation.

But it's still treating this stanza structure as a direct invite and not a mediated invite. Still figuring out how best to fix that part.

**Comment by marc0s** [comment]
As an informational post...

I was trying to find where the user generating the invite was building that stanza, but it doesn't come from the client itself.

Seems that prosody is including that `jabber:x:conference` bit for old clients' compat:

```-- COMPAT: Some older clients expect this
module:hook("muc-invite", function(event)
	local room, stanza = event.room, event.stanza;
	local invite = stanza:get_child("x", "http://jabber.org/protocol/muc#user"):get_child("invite");
	local reason = invite:get_child_text("reason");
	stanza:tag('x', {xmlns = "jabber:x:conference"; jid = room.jid;})
		:text(reason or "")
	:up();
end);```

from `muc.lib.lua` in prosody's 0.11 tree.

**Comment by legastero** [comment]
This should be fully fixed now in 12.3.3.

Thank you for the report!
